# **Menginstall Libraries**

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

# **Mengunduh Data**

In [56]:
data_path = 'https://raw.githubusercontent.com/IBM/telco-customer-churn-on-icp4d/refs/heads/master/data/Telco-Customer-Churn.csv'
df = pd.read_csv(data_path)

further infos: https://community.ibm.com/community/user/businessanalytics/blogs/steven-macko/2019/07/11/telco-customer-churn-1113

penjelasan dataset

# **Memahami & Menilai Data**

Check data summary

In [10]:
# shape dari data
df.shape

(7043, 21)

In [11]:
# 5 baris pertama data
df.head().T

0             1               2  \
customerID              7590-VHVEG    5575-GNVDE      3668-QPYBK   
gender                      Female          Male            Male   
SeniorCitizen                    0             0               0   
Partner                        Yes            No              No   
Dependents                      No            No              No   
tenure                           1            34               2   
PhoneService                    No           Yes             Yes   
MultipleLines     No phone service            No              No   
InternetService                DSL           DSL             DSL   
OnlineSecurity                  No           Yes             Yes   
OnlineBackup                   Yes            No             Yes   
DeviceProtection                No           Yes              No   
TechSupport                     No            No              No   
StreamingTV                     No            No              No   
StreamingMovies                 No            No              No   
Contract            Month-to-month      One year  Month-to-month   
PaperlessBilling               Yes            No             Yes   
PaymentMethod     Electronic check  Mailed check    Mailed check   
MonthlyCharges               29.85         56.95           53.85   
TotalCharges                 29.85        1889.5          108.15   
Churn                           No            No             Yes   

                                          3                 4  
customerID                       7795-CFOCW        9237-HQITU  
gender                                 Male            Female  
SeniorCitizen                             0                 0  
Partner                                  No                No  
Dependents                               No                No  
tenure                                   45                 2  
PhoneService                             No               Yes  
MultipleLines              No phone service                No  
InternetService                         DSL       Fiber optic  
OnlineSecurity                          Yes                No  
OnlineBackup                             No                No  
DeviceProtection                        Yes                No  
TechSupport                             Yes                No  
StreamingTV                              No                No  
StreamingMovies                          No                No  
Contract                           One year    Month-to-month  
PaperlessBilling                         No               Yes  
PaymentMethod     Bank transfer (automatic)  Electronic check  
MonthlyCharges                         42.3              70.7  
TotalCharges                        1840.75            151.65  
Churn                                    No               Yes

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [19]:
# menghitung missing values
df.isna().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [18]:
# deskripsi statistik data
df.describe(include='all').T

count unique               top  freq       mean        std  \
customerID          7043   7043        3186-AJIEK     1        NaN        NaN   
gender              7043      2              Male  3555        NaN        NaN   
SeniorCitizen     7043.0    NaN               NaN   NaN   0.162147   0.368612   
Partner             7043      2                No  3641        NaN        NaN   
Dependents          7043      2                No  4933        NaN        NaN   
tenure            7043.0    NaN               NaN   NaN  32.371149  24.559481   
PhoneService        7043      2               Yes  6361        NaN        NaN   
MultipleLines       7043      3                No  3390        NaN        NaN   
InternetService     7043      3       Fiber optic  3096        NaN        NaN   
OnlineSecurity      7043      3                No  3498        NaN        NaN   
OnlineBackup        7043      3                No  3088        NaN        NaN   
DeviceProtection    7043      3                No  3095        NaN        NaN   
TechSupport         7043      3                No  3473        NaN        NaN   
StreamingTV         7043      3                No  2810        NaN        NaN   
StreamingMovies     7043      3                No  2785        NaN        NaN   
Contract            7043      3    Month-to-month  3875        NaN        NaN   
PaperlessBilling    7043      2               Yes  4171        NaN        NaN   
PaymentMethod       7043      4  Electronic check  2365        NaN        NaN   
MonthlyCharges    7043.0    NaN               NaN   NaN  64.761692  30.090047   
TotalCharges        7043   6531              20.2    11        NaN        NaN   
Churn               7043      2                No  5174        NaN        NaN   

                    min   25%    50%    75%     max  
customerID          NaN   NaN    NaN    NaN     NaN  
gender              NaN   NaN    NaN    NaN     NaN  
SeniorCitizen       0.0   0.0    0.0    0.0     1.0  
Partner             NaN   NaN    NaN    NaN     NaN  
Dependents          NaN   NaN    NaN    NaN     NaN  
tenure              0.0   9.0   29.0   55.0    72.0  
PhoneService        NaN   NaN    NaN    NaN     NaN  
MultipleLines       NaN   NaN    NaN    NaN     NaN  
InternetService     NaN   NaN    NaN    NaN     NaN  
OnlineSecurity      NaN   NaN    NaN    NaN     NaN  
OnlineBackup        NaN   NaN    NaN    NaN     NaN  
DeviceProtection    NaN   NaN    NaN    NaN     NaN  
TechSupport         NaN   NaN    NaN    NaN     NaN  
StreamingTV         NaN   NaN    NaN    NaN     NaN  
StreamingMovies     NaN   NaN    NaN    NaN     NaN  
Contract            NaN   NaN    NaN    NaN     NaN  
PaperlessBilling    NaN   NaN    NaN    NaN     NaN  
PaymentMethod       NaN   NaN    NaN    NaN     NaN  
MonthlyCharges    18.25  35.5  70.35  89.85  118.75  
TotalCharges        NaN   NaN    NaN    NaN     NaN  
Churn               NaN   NaN    NaN    NaN     NaN

In [20]:
print(f'banyaknya data yang terduplikasi: {df.duplicated().sum()}')

banyaknya data yang terduplikasi: 0


Ada kemungkinan salah penulisan atau double writting dengan makna yang sama, seperti: yes, Yess, Yes, sehingga perlu memastikan unique values yang ada.

In [23]:
# cek incorrect values
for col in df.select_dtypes(include=['object']).columns:
    print(df[col].value_counts(dropna=False))
    print("-" * 20)

customerID
3186-AJIEK    1
7590-VHVEG    1
5575-GNVDE    1
8775-CEBBJ    1
2823-LKABH    1
             ..
6713-OKOMC    1
1452-KIOVK    1
9305-CDSKC    1
9237-HQITU    1
7795-CFOCW    1
Name: count, Length: 7043, dtype: int64
--------------------
gender
Male      3555
Female    3488
Name: count, dtype: int64
--------------------
Partner
No     3641
Yes    3402
Name: count, dtype: int64
--------------------
Dependents
No     4933
Yes    2110
Name: count, dtype: int64
--------------------
PhoneService
Yes    6361
No      682
Name: count, dtype: int64
--------------------
MultipleLines
No                  3390
Yes                 2971
No phone service     682
Name: count, dtype: int64
--------------------
InternetService
Fiber optic    3096
DSL            2421
No             1526
Name: count, dtype: int64
--------------------
OnlineSecurity
No                     3498
Yes                    2019
No internet service    1526
Name: count, dtype: int64
--------------------
OnlineBackup
No   

## **Insights**:

#### **1. Data Types & Non-null Counts:**
* Dataset berisi **7043** baris dan **21** kolom.
* **Data types**: Terdapat 18 kolom dengan tipe data `object`, 2 kolom dengan tipe data `int64`, dan 1 kolom dengan tipe data `float64`.
     * **Numerical Features**: Features yang termasuk adalah `tenure` dan `MonthlyCharges`.
     * **Categorical Features**: Features yang termasuk adalah `Partner`, `gender`, etc.

#### **2. Summary Statistics:**
* **Categorical Data**: 
     * terdapat 2/3 data unique
* **Numerical Data**: 
     * `tenure` memiliki rata-rata sebesar `32.37` atau setara 32 hari dari range berlangganan `0` hingga `72` hari. Hal ini menunjukkan sebagian besar pelanggan berhenti lebih awal, yang mengindikasikan perusahaan saat ini belum bisa menjaga hubungan jangka panjang dengan pelanggan dan ini bisa menjadi indikator adanya ketidakpuasan terhadap layanan yang diberikan. 
     * `MonthlyCharges` memiliki rata-rata sebesar `$64.76` dengan standar deviasi sebesar `$30.09`, mengindikasikan variasi harga pembayaran bulanan antar pelanggan yang beragam. Rentang transaksi bulanan minimal `$18.25` dan maksimal `$118.75`, menunjukkan perbedaan yang sangat signifikan dalam transaksi dan terdapat tendensi untuk memaksimalkan penggunaan layanan yang disediakan oleh perusahaan kepada pelanggan.

#### **3. Missing Values Summary:**
* Untuk saat ini tidak terdapat kolom yang berisi missing values.

#### **4. Duplicate & Unique Values Summary:**
* Dari `df.duplicated()`, unique values, dan `df.head()` tidak terlihat ada data yang terduplikasi.
* `customerID` tidak diperlukan untuk proses modeling.
* `TotalCharges` tidak tepat menggunakan tipe data object.
* Tidak ada unique values yang berasal dari salah penulisan atau double writting.

#### **Masalah yang Bisa Diselesaikan:**
* Penulisan nama kolom tidak konsisten, seperti: lower case, diawali huruf kecil/kapital, dan camel case.
* Penulisan kapitalisasi data dalam setiap baris tidak konsisten dan nama kolom juga, sehingga bisa diubah menjadi lower case semua.
* Kolom `SeniorCitizen` berisi data kategorikal apakah seseorang merupakan warga lanjut usia atau tidak dan bukan data untuk perhitungan matematis jadi data type-nya bisa diubah dari `int64` menjadi `object`.
* Kolom `TotalCharges` berisi data numerik kontinu sehingga data type-nya perlu diubah, dari `object` menjadi `float64`.


# **Data Cleaning**

#### Column Renaming & Data Consistency

Penamaan kolom tidak konsisten sehingga perlu dirubah agar konsisten dan sesuai standar.

In [57]:
df.columns = df.columns.str.lower()

Data dalam setiap baris memiliki format penulisan yang tidak konsisten sehingga perlu diubah juga.

In [58]:
categorical_columns = list(df.columns[df.dtypes == 'object'])

In [59]:
for col in categorical_columns:
     df[col] = df[col].str.lower()

#### Correcting Data Type 

Mengubah tipe data dalam setiap kolom yang tidak sesuai dalam dataset.

In [62]:
df.totalcharges = pd.to_numeric(df.totalcharges, errors='coerce')

**Insights:**
* mengubah tipe data object menjadi numerik yang sesuai, dalam konteks kolom ini maka tipe data berubah dari `object` menjadi `float64`.
* `errors='coerce'` nilai yang tidak valid atau tidak dapat diubah ke tipe data numerik akan otomatis digantikan dengan `NaN`.
* karena sebelumnya missing values 0 maka saat ini kemungkinan akan ada missing values di kolom `totalcharges` dan perlu di cleaning.

#### Handling Missing Values

menangani missing values berdasarkan data dan bias

In [63]:
df.isna().sum()

customerid           0
gender               0
seniorcitizen        0
partner              0
dependents           0
tenure               0
phoneservice         0
multiplelines        0
internetservice      0
onlinesecurity       0
onlinebackup         0
deviceprotection     0
techsupport          0
streamingtv          0
streamingmovies      0
contract             0
paperlessbilling     0
paymentmethod        0
monthlycharges       0
totalcharges        11
churn                0
dtype: int64

In [64]:
df[df.totalcharges.isna()].T

488                  753   \
customerid                       4472-lvygi           3115-czmzd   
gender                               female                 male   
seniorcitizen                             0                    0   
partner                                 yes                   no   
dependents                              yes                  yes   
tenure                                    0                    0   
phoneservice                             no                  yes   
multiplelines              no phone service                   no   
internetservice                         dsl                   no   
onlinesecurity                          yes  no internet service   
onlinebackup                             no  no internet service   
deviceprotection                        yes  no internet service   
techsupport                             yes  no internet service   
streamingtv                             yes  no internet service   
streamingmovies                          no  no internet service   
contract                           two year             two year   
paperlessbilling                        yes                   no   
paymentmethod     bank transfer (automatic)         mailed check   
monthlycharges                        52.55                20.25   
totalcharges                            NaN                  NaN   
churn                                    no                   no   

                          936                  1082                     1340  \
customerid          5709-lvoeq           4367-nuyao               1371-dwpaz   
gender                  female                 male                   female   
seniorcitizen                0                    0                        0   
partner                    yes                  yes                      yes   
dependents                 yes                  yes                      yes   
tenure                       0                    0                        0   
phoneservice               yes                  yes                       no   
multiplelines               no                  yes         no phone service   
internetservice            dsl                   no                      dsl   
onlinesecurity             yes  no internet service                      yes   
onlinebackup               yes  no internet service                      yes   
deviceprotection           yes  no internet service                      yes   
techsupport                 no  no internet service                      yes   
streamingtv                yes  no internet service                      yes   
streamingmovies            yes  no internet service                       no   
contract              two year             two year                 two year   
paperlessbilling            no                   no                       no   
paymentmethod     mailed check         mailed check  credit card (automatic)   
monthlycharges           80.85                25.75                    56.05   
totalcharges               NaN                  NaN                      NaN   
churn                       no                   no                       no   

                                 3331                 3826  \
customerid                 7644-omvmy           3213-vvolg   
gender                           male                 male   
seniorcitizen                       0                    0   
partner                           yes                  yes   
dependents                        yes                  yes   
tenure                              0                    0   
phoneservice                      yes                  yes   
multiplelines                      no                  yes   
internetservice                    no                   no   
onlinesecurity    no internet service  no internet service   
onlinebackup      no internet service  no internet service   
deviceprotection  no internet service

In [65]:
missing_ratio = (df['totalcharges'].isna().sum() / len(df)) * 100
print(f"Missing ratio: {missing_ratio:.2f}%")

Missing ratio: 0.16%


**Insights**:
* Terdapat missing values di kolom `totalcharges` sebanyak 11.
* Terlihat data yang memiliki missing values bernilai `no` di kolom `churn` yang artinya pelanggan tidak memiliki kecenderungan atau karakteristik untuk berhenti.
* Meskipun pelanggan tidak churn, untuk menentukan langkah selanjutnya (di-drop atau dipertahankan), perlu memperhatikan bias yang mungkin ada:
     * Kolom `monthlycharges` sudah berisi harga tagihan bulanan.
     * Namun, semua kolom `tenure` berisi angka `0`, hal ini mengindikasikan bahwa 11 orang ini adalah pelanggan baru.
     * Data dari pelanggan baru tidak memiliki kontribusi yang signifikan, tidak berisi data maupun pola historis yang diperlukan. 
     
* Proporsi missing values ini sangat kecil dibandingkan total data (11 dari 7043), yaitu sekitar 0.16%

**Conlusion**: 11 data ini bisa di-drop karena tidak berkontribusi signifikan untuk model.


In [66]:
# drop missing values
df = df.dropna(subset=['totalcharges'])

# **Exploratory Data Analysis**

In [68]:
df.describe(include='all')

customerid gender  seniorcitizen partner dependents       tenure  \
count         7032   7032    7032.000000    7032       7032  7032.000000   
unique        7032      2            NaN       2          2          NaN   
top     3186-ajiek   male            NaN      no         no          NaN   
freq             1   3549            NaN    3639       4933          NaN   
mean           NaN    NaN       0.162400     NaN        NaN    32.421786   
std            NaN    NaN       0.368844     NaN        NaN    24.545260   
min            NaN    NaN       0.000000     NaN        NaN     1.000000   
25%            NaN    NaN       0.000000     NaN        NaN     9.000000   
50%            NaN    NaN       0.000000     NaN        NaN    29.000000   
75%            NaN    NaN       0.000000     NaN        NaN    55.000000   
max            NaN    NaN       1.000000     NaN        NaN    72.000000   

       phoneservice multiplelines internetservice onlinesecurity  ...  \
count          7032          7032            7032           7032  ...   
unique            2             3               3              3  ...   
top             yes            no     fiber optic             no  ...   
freq           6352          3385            3096           3497  ...   
mean            NaN           NaN             NaN            NaN  ...   
std             NaN           NaN             NaN            NaN  ...   
min             NaN           NaN             NaN            NaN  ...   
25%             NaN           NaN             NaN            NaN  ...   
50%             NaN           NaN             NaN            NaN  ...   
75%             NaN           NaN             NaN            NaN  ...   
max             NaN           NaN             NaN            NaN  ...   

       deviceprotection techsupport streamingtv streamingmovies  \
count              7032        7032        7032            7032   
unique                3           3           3               3   
top                  no          no          no              no   
freq               3094        3472        2809            2781   
mean                NaN         NaN         NaN             NaN   
std                 NaN         NaN         NaN             NaN   
min                 NaN         NaN         NaN             NaN   
25%                 NaN         NaN         NaN             NaN   
50%                 NaN         NaN         NaN             NaN   
75%                 NaN         NaN         NaN             NaN   
max                 NaN         NaN         NaN             NaN   

              contract paperlessbilling     paymentmethod monthlycharges  \
count             7032             7032              7032    7032.000000   
unique               3                2                 4            NaN   
top     month-to-month              yes  electronic check            NaN   
freq              3875             4168              2365            NaN   
mean               NaN              NaN               NaN      64.798208   
std                NaN              NaN               NaN      30.085974   
min                NaN              NaN               NaN      18.250000   
25%                NaN              NaN               NaN      35.587500   
50%                NaN              NaN               NaN      70.350000   
75%                NaN              NaN               NaN      89.862500   
max                NaN              NaN               NaN     118.750000   

        totalcharges  churn  
count    7032.000000   7032  
unique           NaN      2  
top              NaN     no  
freq             NaN   5163  
mean     2283.300441    NaN  
std      2266.771362    NaN  
min        18.800000    NaN  
25%       401.450000    NaN  
50%      1397.475000    NaN  
75%      3794.737500    NaN  
max      8684.800000    NaN  

[11 rows x 21 columns]

# **Data Preparation**

# **Modeling**

# **Evaluation**